src/algorithms/perlinnoise.jl

```diff
@@ -0,0 +1,116 @@
+"""
+    PerlinNoise
+    PerlinNoise(; kw...)
+
+Create a Perlin noise neutral landscape model with values ranging 0-1.
+
+# Keywords
+
+- `periods::Tuple{Int,Int}`: the number of periods of Perlin noise across row and 
+    column dimensions for the first octave.
+- `octaves::Int`: the number of octaves that will form the Perlin noise.
+- `lacunarity::Int` : the rate at which the frequency of periods increases for each octive.
+- `persistance::Real` : The rate at which the amplitude of periods decreases for each octive.
+- `valley::String`: The kind of valley bottom that will be mimicked: `:u` (the defualt) 
+    produces u-shaped valleys, `:v` produces v-shaped valleys, and `:-` 
+    produces flat bottomed valleys.
+"""
+Base.@kwdef struct PerlinNoise{PT<:Real} <: NeutralLandscapeMaker 
+    periods::Tuple{Int,Int}
+    octaves::Int = 1
+    lacunarity::Int = 2
+    persistance::PT = 0.5
+    valley::Symbol = :u
+end
+
+function _landscape!(A, alg::PerlinNoise)
+    # nRow must equal nCol so determine the dimension of the smallest square
+    dim = max(size(A)...)
+    # Check the dim is a multiple of each octives maximum number of periods and 
+    # expand dim if needed
+    rperiodsmax = alg.periods[1] * alg.lacunarity^(alg.octaves - 1)
+    cperiodsmax = alg.periods[2] * alg.lacunarity^(alg.octaves - 1)
+    periodsmultiple = lcm(rperiodsmax, cperiodsmax) # lowest commom multiple
+    if dim % periodsmultiple != 0
+        dim = ceil(Int, dim / periodsmultiple) * periodsmultiple
+    end
+
+    # Generate the Perlin noise
+    noise = zeros((dim, dim))
+    for octave in 1:alg.octaves
+        noise .+= octave_noise(alg, octave, dim, dim)
+    end
+
+    # If needed randomly extract the desired array size
+    if size(A) != size(noise)
+        noise = extract_from_square(noise, size(A)...)    
+    end
+    
+    # Rescale the Perlin noise to mimic different kinds of valley bottoms
+    return if alg.valley == :u
+        noise
+    elseif alg.valley == :v
+        abs(noise)
+    elseif alg.valley == :-
+        noise ^ 2
+    end
+end
+
+# This is probably wrong, need to test agains numpy.mgrid
+meshgrid(x, y) = [i for i in x, j in 1:length(y)], [j for i in 1:length(x), j in y]
+
+function octave_noise(alg::PerlinNoise, octave, nrow, ncol)        
+    f(t) = 6 * t ^ 5 - 15 * t ^ 4 + 10 * t ^ 3 # Wut
```

Comment:
would be good with a more informative comment :joy:

Reply:
thats literally the function from nlmpy, but they have less comments

Reply:
haha

Reply:
Still more understandable than [the 2002 original in Java](https://mrl.cs.nyu.edu/~perlin/noise/)

Reply:
he really went for the magic numbers. But that implementation looks fast, I should have read that first